crates/sage/src/ml/mobility_model.rs

```diff
@@ -0,0 +1,296 @@
+const PEPTIDE_MZ: usize = FEATURES - 4;
+const PEPTIDE_LEN: usize = FEATURES - 3;
+const PEPTIDE_MASS: usize = FEATURES - 2;
+const INTERCEPT: usize = FEATURES - 1;
+
+// IN THEORY we could have only one model for both RT and IM
+// And the RT should ignore the charge state "at training time"
+impl MobilityModel {
+    /// One-hot encoding of peptide sequences into feature vector
+    /// Note that this currently does not take into account any modifications
+    fn embed(peptide: &Peptide, charge: &u8, map: &[usize; 26]) -> [f64; FEATURES] {
+        let mut embedding = [0.0; FEATURES];
+        let cterm = peptide.sequence.len().saturating_sub(3);
+        let pep_length = peptide.sequence.len() as f64;
+
+        //let default_first_val = 1.0f64;
+        for (aa_idx, residue) in peptide.sequence.iter().enumerate() {
+            let idx = map[(residue - b'A') as usize];
+            embedding[idx] += 1.0;
+            // Embed N- and C-terminal AA's
+            // 2 on each end
+            match aa_idx {
+                0 | 1=> embedding[N_TERMINAL + idx] += 1.0,
+                x if x > cterm => embedding[C_TERMINAL + idx] += 1.0,
+                _ => {}
+            }
+            let x = idx;
+
+            if BULKY_AA_IDXS.contains(&x) {
+                embedding[NUM_BULKY] += 1.0;
+            };
+            if UNCHARGED_POLAR_AA_IDXS.contains(&x) {
+                embedding[NUM_UC_POLAR] += 1.0;
+            };
+            if POSITIVE_AA_IDXS.contains(&x) {
+                embedding[NUM_POSITIVE] += 1.0
+            };
+            if NEGATIVE_AA_IDXS.contains(&x) {
+                embedding[NUM_NEGATIVE] += 1.0
+            };
+            if TINY_AA_IDXS.contains(&x) {
+                embedding[NUM_TINY] += 1.0
+            };
+            if BRANCHED_AA_IDXS.contains(&x) {
+                embedding[NUM_BRANCHED] += 1.0
+            };
+        }
+
+        // PCT features are just the AA counts divided by the length of the peptide
+        for idx in 0..VALID_AA.len() {
+            let pct_val = embedding[idx] / pep_length;
+            embedding[PCT_FEATURES_START + idx] = pct_val;
+        }
+
+        let charge_feature: f64 = *charge as f64;
+        embedding[PEPTIDE_CHARGE] = charge_feature;
+        embedding[INV_PEPTIDE_CHARGE] = 1. / charge_feature;
+        embedding[PEPTIDE_LEN] = peptide.sequence.len() as f64;
+        embedding[PEPTIDE_MASS] = (peptide.monoisotopic as f64) / 1000.0;
+        embedding[PEPTIDE_MZ] = ((peptide.monoisotopic as f64) / charge_feature) / 1000.0;
+        embedding[INTERCEPT] = 1.0;
+        embedding
+    }
+
+    /// Attempt to fit a linear regression model: peptide sequence + charge ~ retention time
+    pub fn fit(db: &IndexedDatabase, training_set: &[Feature]) -> Option<Self> {
+        // Create a mapping from amino acid character to vector embedding
+        // Q: Why has this been implemented as a this map and not a hashmap[byte -> usize]?
```

Comment:
Why implement it as a hashmap? Seems wasteful 😃 - we know that there are only ~20 possible entries - using a lookup table will be probably 100-1000x faster (not that this is a remotely close to being a bottleneck in the program). 

As a general note, Rust's default HashMap is quite slow - it's designed to be DoS resistant/secure. You'll note that hashmaps are used very sparingly in Sage, and when they *are* used, it's typically with an alternative hashing function (FNV), since Sage doesn't really need denial-of-service resistance.

Reply:
That is very informative! thanks a lot! (removed in 8c2ac76)